Create agenda topics

tsiegleauq opened the pull request:
Also:
Enhanced DragNDrop Component
List Of Speakers

FinnStutzenstein commented on client/src/app/core/services/http.service.ts:
```diff
@@ -90,16 +90,21 @@ export class HttpService {
      * @return Observable of object
      */
-    public delete(url: string): Observable<object> {
+    public delete(url: string, obj?: object): Observable<object> {
```
please rebase on the current master ;)

FinnStutzenstein commented on client/src/app/shared/components/sorting-list/sorting-list.component.ts:
```diff
@@ -27,27 +29,56 @@ import { EmptySelectable } from '../empty-selectable';
+    public templateRef: TemplateRef<Selectable>;
+
+    /**
+     * Determine whether changes to the list occur directly after drag'n'drop
```
changes to what: the sortEvent or the listOfSeelctables?

tsiegleauq replied:
What I tried to express was:
if live: sort events should immediately fire events (i.e in list of speakers)
if not: event should be fired after the user hits a save-button (i.e sort motions in categories)

FinnStutzenstein commented on client/src/app/shared/components/sorting-list/sorting-list.component.ts:
```diff
@@ -15,7 +15,9 @@ import { EmptySelectable } from '../empty-selectable';
  * <os-sorting-list
- *   [input]="listOfSelectables">
+ *     [input]="listOfSelectables"
+ *     [live]="true"
```
[count]="true"

FinnStutzenstein commented on client/src/app/site/agenda/components/agenda-list/agenda-list.component.ts:
```diff
@@ -60,6 +65,7 @@ export class AgendaListComponent extends ListViewBaseComponent<ViewItem> impleme
      */
     public onPlusButton(): void {
+        this.router.navigate(['./new'], { relativeTo: this.route });
         console.log('create new motion');
```
please remove

FinnStutzenstein commented on client/src/app/site/agenda/components/speaker-list/speaker-list.component.html:
```diff
@@ -0,0 +1,120 @@
+    <div class="title-slot">
+        <h2>
+            <span *ngIf="viewItem">{{ viewItem.getTitle() }}</span>
+            <span>:</span>
```
this does not need an own span. Just append `:` to the line above.

FinnStutzenstein commented on client/src/app/site/agenda/components/speaker-list/speaker-list.component.html:
```diff
@@ -0,0 +1,120 @@
+                    <!-- <span>&nbsp;</span> -->
+                    <!-- <span translate>minutes</span> -->
+                    <span>&nbsp;(</span>
+                    <span translate>Start time:</span>
```
move the colon out of the span

FinnStutzenstein commented on client/src/app/site/agenda/components/speaker-list/speaker-list.component.ts:
```diff
@@ -0,0 +1,162 @@
+    /**
+     * Hold the users
+     */
+    public users: BehaviorSubject<Array<User>>;
```
`BehaviorSubject<User[]>` for consistency

FinnStutzenstein commented on client/src/app/site/agenda/components/speaker-list/speaker-list.component.ts:
```diff
@@ -0,0 +1,162 @@
+
+    /**
+     * Create a speaker out of an id
+     * @param userIds Array of userIds to add as AgendaItemSpeaker
```
the comment does not match. What happens, if no userId is given?

FinnStutzenstein commented on client/src/app/site/agenda/components/topic-detail/topic-detail.component.html:
```diff
@@ -0,0 +1,80 @@
+
+            <div>
+                <mat-form-field>
+                    <textarea matInput formControlName="text" placeholder="{{ 'Description ' | translate}}">
```
This formatting is strange.

FinnStutzenstein commented on client/src/app/site/agenda/services/agenda-repository.service.ts:
```diff
@@ -45,27 +49,120 @@ export class AgendaRepositoryService extends BaseRepository<ViewItem, Item> {
+     * @param id {@link User} id of the new speaker
+     * @param agenda the target agenda item
+     */
+    public addSpeaker(id: number, agenda: Item): Observable<any> {
```
Change to promise. Same for the complete repository (rebase to master)

FinnStutzenstein commented on client/src/app/site/agenda/services/agenda-repository.service.ts:
```diff
@@ -45,27 +49,120 @@ export class AgendaRepositoryService extends BaseRepository<ViewItem, Item> {
     /**
      * @ignore
-     *
-     * TODO: used over not-yet-existing detail view
```
Do not delete this, if it's not implemented ;)

FinnStutzenstein replied:
Ok, add a comment then, that they should not be used (and may just throw an error there, like in the config repo).

FinnStutzenstein commented on client/src/app/site/agenda/components/topic-detail/topic-detail.component.html:
```diff
@@ -0,0 +1,80 @@
+        <div>
+            <span *ngIf=!editTopic>
+                {{topic.text}}
+            </span>
```
Make some gray notice, if there is no text, so there isn't a empty white box. Maybe the same as in statute paragraphs, motion comments or motion comment sections.

FinnStutzenstein commented:
What do you think of making the mad-card (e.g. in the list of speakers) full width (with a margin at the right like left and a margin on top)? Currently this seems a bit strange with so much gray right. But put around the content a div with `max-width: 1200px`, so the content is still limited in width.

The last, current and waiting speakers works fine. Add/remove users (incl. the "add me" button) works.

What does not work is d&d. As seen in the screenshot I can drad the entries around and drop them whereever I want. But there is nothing like snapping back into the list and no sorting is done when I release the entries. This needs some more investigations..
![dnd](https://user-images.githubusercontent.com/8439581/47864423-d5e5c300-ddf9-11e8-890a-aad093b16c7d.jpg)

tsiegleauq commented:
> What do you think of making the mad-card (e.g. in the list of speakers) full width (with a margin at the right like left and a margin on top)? Currently this seems a bit strange with so much gray right. But put around the content a div with `max-width: 1200px`, so the content is still limited in width.
> 
> The last, current and waiting speakers works fine. Add/remove users (incl. the "add me" button) works.
> 
> What does not work is d&d. As seen in the screenshot I can drad the entries around and drop them whereever I want. But there is nothing like snapping back into the list and no sorting is done when I release the entries. This needs some more investigations..
> ![dnd](https://user-images.githubusercontent.com/8439581/47864423-d5e5c300-ddf9-11e8-890a-aad093b16c7d.jpg)

@FinnStutzenstein 
Drag and Drop should work. Remove your node modules and ensure you updated to angular 7 using npm install.
(furthermore, it looks wrong. I provide a screenshot below)

Limiting the width was requested by our design "team". Add this to #3951 please. We need to limit the with in general I suppose. 

![localhost_4200_agenda_1](https://user-images.githubusercontent.com/10233032/47865774-61148800-ddfd-11e8-82a9-1fb74dd3b91b.png)

FinnStutzenstein commented:
OK, I reinstalled angular and now it works. Then do not do any width limiting and this can be done consistently for all views in a separate PR. what do you think?

tsiegleauq commented:
> OK, I reinstalled angular and now it works. Then do not do any width limiting and this can be done consistently for all views in a separate PR. what do you think?

Indeed - I Can remove that here. I take care about limiting the width for all site/*-pages in another GUI rework.

tsiegleauq commented:
Rebased and reworked.
I noticed some race conditions and made an issue for them. Did not fix them here to not extend the waiting time for the features (and I race conditions are potentially difficult to solve)
@FinnStutzenstein

FinnStutzenstein commented:
@tsiegleauq please resolve the merge conflicts

FinnStutzenstein commented on client/src/app/site/agenda/services/agenda-repository.service.ts:
```diff
@@ -74,7 +164,9 @@ export class AgendaRepositoryService extends BaseRepository<ViewItem, Item> {
     public createViewModel(item: Item): ViewItem {
         const contentObject = this.getContentObject(item);
-
+        if (!contentObject) {
```
Please remove this

tsiegleauq replied:
Removed now. Actually the warnings were supposed to stay until race conditions have been resolved to track them more easily.

FinnStutzenstein commented on client/src/app/site/agenda/services/topic-repository.service.ts:
```diff
@@ -0,0 +1,85 @@
+    public createViewModel(topic: Topic): ViewTopic {
+        const attachments = this.DS.getMany(Mediafile, topic.attachments_id);
+        const item = this.getAgendaItem(topic);
+        if (!item) {
```
Please remove

tsiegleauq commented:
More conflicts. Well well.

FinnStutzenstein commented on client/src/app/site/agenda/components/speaker-list/speaker-list.component.html:
```diff
@@ -0,0 +1,116 @@
+                    matTooltip="{{ 'Remove' | translate }}"
+                    *osPerms="'agenda.can_manage_list_of_speakers'"
+                    (click)="onDeleteButton(speaker)"
+                >
```
Please put closing brackets in the same row as the last attribute. Same for every occurence..

tsiegleauq replied:
That did not work in my IDE somehow. I suppose there was a new package to automatically align HTML

tsiegleauq replied:
tried multiple times with other settings. Apparently prettify supports html now.

jsaalfeld commented on client/src/app/site/agenda/components/topic-detail/topic-detail.component.ts:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    /**
+     * Create the absolute path to the corresponding list of speakers
```
proper `@returns` is missing in the documentation

tsiegleauq replied:
yep

jsaalfeld commented on client/src/app/site/agenda/components/topic-detail/topic-detail.component.ts:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    /**
+     * Hitting escape while in topicForm should cancel editing
```
the `@param` is missing in documentation

tsiegleauq replied:
yep

jsaalfeld commented on client/src/app/site/agenda/models/view-speaker.ts:
```diff
@@ -0,0 +1,86 @@
+        return this.speaker ? this.speaker.end_time : null;
+    }
+
+    public get state(): SpeakerState {
```
Since this function has a bit logic, a documentation would be nice.

tsiegleauq replied:
In that case yes.

jsaalfeld commented on client/src/app/site/agenda/models/view-speaker.ts:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    /**
+     * Using short name for now. Could be full name.
```
proper `@returns` is missing in documentation

tsiegleauq replied:
Comment should be removed since it's wrong (do you even read) ;-)

jsaalfeld commented on client/src/app/site/agenda/models/view-speaker.ts:
```diff
@@ -0,0 +1,86 @@
+        return this.user.full_name;
+    }
+
+    public constructor(speaker?: Speaker, user?: User) {
```
function documentation is missing

tsiegleauq replied:
that's a model. Models currently have no comments.

jsaalfeld commented on client/src/app/site/agenda/models/view-topic.ts:
```diff
@@ -0,0 +1,58 @@
+        return this.topic ? this.topic.text : null;
+    }
+
+    public constructor(topic?: Topic, attachments?: Mediafile[], item?: Item) {
```
documentation is missing

jsaalfeld commented on client/src/app/site/agenda/models/view-topic.ts:
```diff
@@ -0,0 +1,58 @@
+        return this.title;
+    }
+
+    public updateValues(update: Topic): void {
```
documentation is missing

jsaalfeld commented on client/src/app/site/agenda/services/agenda-repository.service.ts:
```diff
@@ -18,15 +22,20 @@ import { CollectionStringModelMapperService } from '../../../core/services/colle
 })
 export class AgendaRepositoryService extends BaseRepository<ViewItem, Item> {
-    public constructor(DS: DataStoreService, mapperService: CollectionStringModelMapperService) {
+    public constructor(
```
function documentation is missing

tsiegleauq replied:
yep

jsaalfeld commented on client/src/app/site/agenda/services/agenda-repository.service.ts:
```diff
@@ -18,15 +22,20 @@ import { CollectionStringModelMapperService } from '../../../core/services/colle
     /**
      * Returns the corresponding content object to a given {@link Item} as an {@link AgendaBaseModel}
      * @param agendaItem
+     * Used dynamically because of heavy race conditions
```
proper `@returns` is missing in the documentation

tsiegleauq replied:
yep

jsaalfeld commented on client/src/app/site/agenda/services/agenda-repository.service.ts:
```diff
@@ -45,6 +54,87 @@ export class AgendaRepositoryService extends BaseRepository<ViewItem, Item> {
 
+    /**
+     * Generate viewSpeaker objects from a given agenda Item
+     * @param item agenda Item holding speakers
```
proper `@returns` is missing in the documentation

tsiegleauq replied:
yep

jsaalfeld commented on client/src/app/site/agenda/services/agenda-repository.service.ts:
```diff
@@ -45,6 +54,87 @@ export class AgendaRepositoryService extends BaseRepository<ViewItem, Item> {
+    /**
+     * Marks the current speaker
+     * @param id {@link User} id of the new speaker
+     * @param doMark determine if the user was marked or not
```
`doMark` was renamed to `mark`

tsiegleauq replied:
yep. thanks!

jsaalfeld commented on client/src/app/site/agenda/services/agenda-repository.service.ts:
```diff
@@ -74,7 +164,9 @@ export class AgendaRepositoryService extends BaseRepository<ViewItem, Item> {
 
     public createViewModel(item: Item): ViewItem {
```
documentation is missing

jsaalfeld commented on client/src/app/site/agenda/components/agenda-list/agenda-list.component.ts:
```diff
@@ -51,20 +53,20 @@ export class AgendaListComponent extends ListViewBaseComponent<ViewItem> impleme
      * Links to the content object if any
+     *
+     * Gets content object from the repository rather than from the model
+     * to avoid race conditions
```
the `@param` is missing in documentation

jsaalfeld commented on client/src/app/site/agenda/components/speaker-list/speaker-list.component.ts:
```diff
@@ -0,0 +1,160 @@
+
+    /**
+     * Click on the mic-cross button
+     * @param item
```
the param is outdated

jsaalfeld commented on client/src/app/site/agenda/services/topic-repository.service.ts:
```diff
@@ -0,0 +1,82 @@
+     * Constructor calls the parent constructor
+     * @param DS Access the DataStore
+     * @param dataSend Access the DataSendService
+     */
```
the `mapperService` is missing in the parameters

jsaalfeld commented on client/src/app/site/agenda/services/topic-repository.service.ts:
```diff
@@ -0,0 +1,82 @@
+
+    /**
+     * Gets the corresponding agendaItem to the topic.
+     * Used to deal with race conditions
```
proper `@param` and `@returns` are missing from the documentation

jsaalfeld commented on client/src/app/site/agenda/services/topic-repository.service.ts:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * Creates a new viewModel out of the given model
+     * @param topic The topic that shall be converted into a view topic
+     */
```
proper `@returns` is missing in thedocumentionation

jsaalfeld commented on client/src/app/site/agenda/services/topic-repository.service.ts:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * Save a new topic
+     * @param topicData Partial topic data to be created
+     */
```
proper `@returns` is missing in the documentation

jsaalfeld commented on client/src/app/site/motions/components/motion-detail/motion-detail.component.ts:
```diff
@@ -322,17 +322,17 @@ export class MotionDetailComponent extends BaseViewComponent implements OnInit {
-            motList.splice(index, 1);
-        }
-        this.categoryRepo.updateCategoryNumbering(this.motion.category, motList);*/
+    public async deleteMotionButton(): Promise<void> {
```
Update the documentation accordingly

jsaalfeld commented on client/src/app/site/motions/components/motion-detail/motion-detail.component.ts:
```diff
@@ -492,6 +491,13 @@ export class MotionDetailComponent extends BaseViewComponent implements OnInit {
 
+    /**
+     * Create the absolute path to the corresponding list of speakers
+     */
```
proper `@returns` is missing from the documentation

tsiegleauq commented:
@FinnStutzenstein done.

jsaalfeld commented on client/src/app/site/agenda/models/view-speaker.ts:
```diff
@@ -0,0 +1,86 @@
+     * Speaker is not a base model,
+     * @param update the incoming update
+     */
+    public updateValues(update: any): void {
```
shouldn't this be a proper class? Like a meta-class, like in the category-numbering?

tsiegleauq replied:
Perhaps it should. The base repository would need to be adjusted for that.
If this was proven to be necessary, create an issue.

jsaalfeld replied:
Noted that in #3999

tsiegleauq commented on client/src/app/shared/components/sorting-list/sorting-list.component.html:
```diff
@@ -1,10 +1,17 @@
         <div class="section-one" cdkDragHandle>
-            <mat-icon>drag_handle</mat-icon>
+            <!-- TODO: drag_indicator after icon update -->
+            <mat-icon>unfold_more</mat-icon>
```
yes.

tsiegleauq commented on client/src/app/core/services/http.service.ts:
```diff
@@ -33,14 +32,14 @@ export class HttpService {
         }
 
         const options = {
```
This is not a change in this pull request, thus not to be changed here.

jsaalfeld replied:
You could change it in this PR nonetheless. It would save some work down the road if we change now everything we see. As something like "good practice". But I understand your point and will accept it.

tsiegleauq commented on client/src/app/shared/components/sorting-list/sorting-list.component.html:
```diff
@@ -1,10 +1,17 @@
         </div>
         <div class="section-two">
-            {{item}}
+            <!-- {number}. {item.toString()} -->
```
Yes.
Usually, HTML comments are quired and this comment helps to tell how the format of the string should look like.

tsiegleauq commented on client/src/app/shared/components/sorting-list/sorting-list.component.ts:
```diff
@@ -27,27 +30,62 @@ import { EmptySelectable } from '../empty-selectable';
      */
     @Input()
-    public input: Array<Selectable>;
+    public live = false;
```
No its not. Look above.

jsaalfeld replied:
Sorry, Misread.

tsiegleauq commented on client/src/app/shared/components/sorting-list/sorting-list.component.ts:
```diff
@@ -56,5 +94,11 @@ export class SortingListComponent implements OnInit {
+    /**
+     * Required by components using the selector as directive
+     */
+    public ngOnInit(): void {}
```
Documentation is above. I move the init some lines up then.

tsiegleauq commented on client/src/app/shared/models/topics/topic.ts:
```diff
@@ -25,6 +25,6 @@ export class Topic extends AgendaBaseModel {
     }
 
     public getDetailStateURL(): string {
```
domain data do not have comments.... that is why there is an ignore above.

tsiegleauq commented on client/src/app/site/agenda/agenda.module.ts:
```diff
@@ -3,10 +3,11 @@ import { CommonModule } from '@angular/common';
-    declarations: [AgendaListComponent]
+    declarations: [AgendaListComponent, TopicDetailComponent]
 })
 export class AgendaModule {}
```
yep

tsiegleauq commented on client/src/app/site/agenda/components/agenda-list/agenda-list.component.ts:
```diff
@@ -25,6 +27,7 @@ export class AgendaListComponent extends ListViewBaseComponent<ViewItem> impleme
     public constructor(
         titleService: Title,
         translate: TranslateService,
+        private route: ActivatedRoute,
```
yep

tsiegleauq commented on client/src/app/site/agenda/components/speaker-list/speaker-list.component.ts:
```diff
@@ -0,0 +1,160 @@
+    templateUrl: './speaker-list.component.html',
+    styleUrls: ['./speaker-list.component.scss']
+})
+export class SpeakerListComponent implements OnInit {
```
yep

tsiegleauq commented on client/src/app/site/agenda/components/speaker-list/speaker-list.component.ts:
```diff
@@ -0,0 +1,160 @@
+     */
+    public addSpeakerForm: FormGroup;
+
+    public constructor(
```
yep

tsiegleauq commented on client/src/app/site/agenda/components/speaker-list/speaker-list.component.ts:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    /**
+     * Returns true if the operator is in the list of (waiting) speakers
```
yep

tsiegleauq commented on client/src/app/site/agenda/components/speaker-list/speaker-list.component.ts:
```diff
@@ -0,0 +1,160 @@
+     * Observe users,
+     * React to form changes
+     */
+    public ngOnInit(): void {
```
yep. So the keep the `ngOn*` stuff now directly under the constructor?
Would you write some guidlines for that please?

jsaalfeld replied:
Sure!

jsaalfeld replied:
Will be maintained at: https://github.com/OpenSlides/OpenSlides/wiki/Typescript-Coding-Guidelines

tsiegleauq commented on client/src/app/site/agenda/components/topic-detail/topic-detail.component.ts:
```diff
@@ -0,0 +1,171 @@
+     */
+    public topicForm: FormGroup;
+
+    public constructor(
```
yep
